R/faceted_qc_plot.R

```diff
@@ -1,84 +1,244 @@
+
+  # If unique and available, format units to a string for plot labels
+  x_unit_lab <- processed_data %>% format_unit_string(x_var_units)
+  colour_unit_lab <- processed_data %>% format_unit_string(colour_var_units)
+
+  # Define a title for the legend
+  legend_title <- paste(
+    paste(
+      ifelse(
+        plot_conc_data,
+        get_label(labels_df, shape_var, "ADPC"),
+        ""
+      ),
+      ifelse(
+        plot_dose_data,
+        paste0(get_label(labels_df, colour_var, "ADPC"), colour_unit_lab),
+        ""
+      ),
+      sep = "<br>"
+    ),
+    "<br>"
+  )
+
+  # Build the plot
+  p <- ggplot(
+    processed_data,
+    aes(
+      x = .data[[x_var]],
+      y = .data[[y_var]],
+      text = tooltip_text,
+      colour = legend_group,
+      shape = legend_group
+    )
+  ) +
+    geom_point(aes(alpha = legend_group), size = 2.5, stroke = 0.25) +
+    facet_wrap(vars(facet_title), scales = "free_y", ncol = 1) +
+
+    # Apply the manual scales
+    scale_shape_manual(name = legend_title, values = shape_values) +
+    scale_colour_manual(name = legend_title, values = colour_values) +
+
+    # Make doses semi-transparent and hide the alpha legend
+    scale_alpha_manual(values = setNames(c(rep(1, length(shape_levels)),
+                                           rep(0.6, length(colour_levels))),
+                                         all_legend_levels), guide = "none") +
     labs(
-      x = get_label(labels_df, x_var, "ADPC"),
+      x = paste0(get_label(labels_df, x_var, "ADPC"), x_unit_lab),
       y = get_label(labels_df, y_var, "ADPC"),
-      title = title,
-      subtitle = paste("Subjects grouped by",
-                       paste(grouping_vars, collapse = ", ")),
-      colour = get_label(labels_df, colour_var, "ADPC")
+      title = title
     ) +
-    theme_bw() +
-    theme(
-      # Keep cohort labels horizontal
-      strip.text.y = element_text(angle = -90),
-      # Adjust spacing between panels
-      panel.spacing = unit(0.2, "lines")
-    )
+    theme_bw()
+
+  if (as_plotly) {
+    ggplotly(p, tooltip = "text") %>%
+      layout(title = list(text = p$labels$title), legend = list(traceorder = "normal"))
+  } else {
+    p
+  }
+}
+
+
+#' Formats a unit string if a unique unit exists
+#' @param data The data frame to check.
+#' @param unit_var The column name of the unit variable.
+#' @return A formatted string like " (hr)" or an empty string "".
+format_unit_string <- function(data, unit_var) {
+  # Return "" if the unit variable is not specified or doesn't exist
+  if (is.null(unit_var) || !all(unit_var %in% names(data))) {
+    return("")
+  }
+
+  # Get the distinct unit values
+  distinct_units <- data %>%
+    select(all_of(unit_var)) %>%
+    distinct()
 
-  plt
+  # If there is exactly one unique unit, format it. Otherwise, return "".
+  if (nrow(distinct_units) == 1) {
+    unit_value <- distinct_units %>% pull()
+    paste0(" (", unit_value, ")")
+  } else {
+    ""
+  }
```

Comment:
nitpick, style: Prefer early return over `if-else`:
```suggestion
  if (nrow(distinct_units) != 1) {
    return("")
  }
  
  distinct_units %>%
    pull()
    paste0(" (", ., ")")
```